core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java

```diff
@@ -0,0 +1,445 @@
+
+/**
+ * Class used to visit all child relations from a root relation and optionally replace subtrees by
+ * overriding a visitor method. The traversal will include relations inside of subquery expressions.
+ * By default, no subtree substitution will be performed. However, if a visit method is overridden
+ * to return a different relation than the one passed in, then that relation will replace it in the
+ * tree.
+ */
+public class RelCopyOnWriteVisitor implements RelVisitor<Rel, RuntimeException> {
+  @Override
+  public Aggregate visit(Aggregate aggregate) {
+    Rel oldInput = aggregate.getInput();
+    Rel newInput = oldInput.accept(this);
+    return oldInput == newInput
+        ? aggregate
+        : ImmutableAggregate.builder().from(aggregate).input(newInput).build();
+  }
+
+  @Override
+  public EmptyScan visit(EmptyScan emptyScan) {
+    return emptyScan;
+  }
+
+  @Override
+  public Fetch visit(Fetch fetch) {
+    Rel oldInput = fetch.getInput();
+    Rel newInput = oldInput.accept(this);
+    return oldInput == newInput
+        ? fetch
+        : ImmutableFetch.builder().from(fetch).input(newInput).build();
+  }
+
+  @Override
+  public Filter visit(Filter filter) {
+    Rel oldInput = filter.getInput();
+    Rel newInput = oldInput.accept(this);
+    Expression oldCondition = filter.getCondition();
+    Expression newCondition = visitExpression(oldCondition);
+    return oldInput == newInput && oldCondition == newCondition
+        ? filter
+        : ImmutableFilter.builder().from(filter).input(newInput).condition(newCondition).build();
+  }
+
+  @Override
+  public Join visit(Join join) {
+    Rel oldLeft = join.getLeft();
+    Rel newLeft = oldLeft.accept(this);
+    Rel oldRight = join.getRight();
+    Rel newRight = oldRight.accept(this);
+    Expression oldCondition = null, newCondition = null;
+    if (join.getCondition().isPresent()) {
+      newCondition = visitExpression(oldCondition = join.getCondition().get());
+    }
+    Expression oldFilter = null, newFilter = null;
+    if (join.getPostJoinFilter().isPresent()) {
+      newFilter = visitExpression(oldFilter = join.getPostJoinFilter().get());
+    }
+    if (oldLeft == newLeft
+        && oldRight == newRight
+        && oldCondition == newCondition
+        && oldFilter == newFilter) {
+      return join;
+    }
+    ImmutableJoin.Builder builder =
+        ImmutableJoin.builder().from(join).left(newLeft).right(newRight);
+    if (join.getCondition().isPresent()) {
+      builder.condition(newCondition);
+    }
+    if (join.getPostJoinFilter().isPresent()) {
+      builder.postJoinFilter(newFilter);
+    }
+    return builder.build();
+  }
+
+  @Override
+  public NamedScan visit(NamedScan namedScan) {
+    return namedScan;
+  }
+
+  private List<Expression> copyOnWrite(List<Expression> oldExpressions) {
+    int expressionCount = oldExpressions.size();
+    List<Expression> newExpressions = oldExpressions;
+    for (int i = 0; i < expressionCount; i++) {
```

Comment:
Rewrite for this function with streams and the optional behavior proposed above:
```
// A utility function to use for optionally transforming a list of things, creating a new list if at least one item has changed.

public static class Pointer<T> {
  public T value;
}

public static <T> Optional<List<T>> transformList(Collection<T> items, Function<T, Optional<T>> transform) {
  var changed = new Pointer<Boolean>();
  var newItems = items.forEach(item => {
    var transformedItem = transform(item);
    if(transformedItem.isPresent()) {
      changed.value = true;
      return transformedItem.get();
    }
    return item;
  });
  if(changed.value) {
   return Optional.of(newItems.toList());
  } 
  return Optional.empty();
}

private Optional<List<Expression>> copyOnWrite(List<Expression> oldExpressions) {
  return transformList(oldExpressions, this::visitExpression);
}
```